src/m3ninx/index/regexp_test.go

```diff
@@ -0,0 +1,284 @@
+	//
+	// We require that '^' and '$' only match input terminating characters (i.e.
+	// text boundaries, not line boundaries within the input). The line of code
+	// below ensures that syntax.Perl does the same.
+	require.NotZero(t, syntax.Perl&syntax.OneLine)
+
+	// ensure our `parseRegexp` internal function uses the right flags too.
+	re, err := parseRegexp(".*")
+	require.NoError(t, err)
+	require.NotZero(t, re.Flags&syntax.OneLine)
+}
+
+func TestEnsureRegexpUnachored(t *testing.T) {
+	testCases := []testCase{
+		testCase{
+			name:           "naked ^",
+			input:          "^",
+			expectedOutput: "(?:)",
+		},
+		testCase{
+			name:           "naked $",
+			input:          "$",
+			expectedOutput: "(?:)",
+		},
+		testCase{
+			name:           "empty string ^$",
+			input:          "^$",
+			expectedOutput: "",
+		},
+		testCase{
+			name:           "invalid naked concat ^$",
+			input:          "$^",
+			expectedOutput: "(?-m:$)\\A",
+		},
+		testCase{
+			name:           "simple case of ^",
+			input:          "^abc",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "simple case of $",
+			input:          "abc$",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "simple case of both ^ & $",
+			input:          "^abc$",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "weird case of internal ^",
+			input:          "^a^bc$",
+			expectedOutput: "a\\Abc",
+		},
+		testCase{
+			name:           "weird case of internal $",
+			input:          "^a$bc$",
+			expectedOutput: "a(?-m:$)bc",
+		},
+		testCase{
+			name:           "alternate of sub expressions with only legal ^ and $",
+			input:          "(?:^abc$)|(?:^xyz$)",
+			expectedOutput: "abc|xyz",
+		},
+		testCase{
+			name:           "concat of sub expressions with only legal ^ and $",
+			input:          "(^abc$)(?:^xyz$)",
+			expectedOutput: "(abc(?-m:$))\\Axyz",
+		},
+		testCase{
+			name:           "alternate of sub expressions with illegal ^ and $",
+			input:          "(?:^a$bc$)|(?:^xyz$)",
+			expectedOutput: "a(?-m:$)bc|xyz",
+		},
+		testCase{
+			name:           "concat of sub expressions with illegal ^ and $",
+			input:          "(?:^a$bc$)(?:^xyz$)",
+			expectedOutput: "a(?-m:$)bc(?-m:$)\\Axyz",
+		},
+		testCase{
+			name:           "question mark case both boundaries success",
+			input:          "(?:^abc$)?",
+			expectedOutput: "(?:abc)?",
+		},
+		testCase{
+			name:           "question mark case only ^",
+			input:          "(?:^abc)?",
+			expectedOutput: "(?:abc)?",
+		},
+		testCase{
+			name:           "question mark case only $",
+			input:          "(?:abc$)?",
+			expectedOutput: "(?:abc)?",
+		},
+		testCase{
+			name:           "question concat case $",
+			input:          "abc$?",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "star mark case both boundaries success",
+			input:          "(?:^abc$)*",
+			expectedOutput: "(?:abc)?(?:\\Aabc(?-m:$))*",
```

Comment:
Hm yeah, what's expected I guess here.  I wonder if `"(?:abc)?"` is more correct translation since we only allow end of input to appear once?